File: src/main/java/mods/eln/sim/mna/component/Line.java
package mods.eln.sim.mna.component;

import mods.eln.sim.mna.RootSystem;
import mods.eln.sim.mna.SubSystem;
import mods.eln.sim.mna.misc.ISubSystemProcessFlush;
import mods.eln.sim.mna.state.State;

import java.util.Iterator;
import java.util.LinkedList;

public class Line extends Resistor implements ISubSystemProcessFlush, IAbstractor {

    public LinkedList<Resistor> resistors = new LinkedList<Resistor>(); //from a to b
    public LinkedList<State> states = new LinkedList<State>(); //from a to b

    boolean ofInterSystem;

    boolean canAddComponent(Component c) {
        return (c instanceof Resistor);
    }

    void addResistor(Resistor c) {
        ofInterSystem |= c.canBeReplacedByInterSystem();
        resistors.add(c);
    }

    @Override
    public boolean canBeReplacedByInterSystem() {
        return ofInterSystem;
    }

    public void recalculateResistance() {
        double resistance = 0;
        for (Resistor r : resistors) {
            resistance += r.getResistance();
        }

        setResistance(resistance);
    }

    void restoreResistorIntoCircuit() {
        this.breakConnection();
    }

    void removeResistorFromCircuit() {
    }

    public static void newLine(RootSystem root, LinkedList<Resistor> resistors, LinkedList<State> states) {
        if (resistors.size() > 1) {
            Resistor first = resistors.getFirst();
            Resistor last = resistors.getLast();
            State stateBefore = first.aPin == states.getFirst() ? first.bPin : first.aPin;
            State stateAfter = last.aPin == states.getLast() ? last.bPin : last.aPin;

            Line line = new Line();
            line.resistors = resistors;
            line.states = states;
            line.recalculateResistance();
            root.addComponents.removeAll(resistors);
            root.addStates.removeAll(states);
            root.addComponents.add(line);
            line.connectTo(stateBefore, stateAfter);
            line.removeResistorFromCircuit();

            root.addProcess(line);

            for (Resistor r : resistors) {
                r.abstractedBy = line;
                line.ofInterSystem |= r.canBeReplacedByInterSystem();
            }

            for (State s : states) {
                s.abstractedBy = line;
            }
        }
    }

    @Override
    public void returnToRootSystem(RootSystem root) {
        for (Resistor r : resistors) {
            r.abstractedBy = null;
        }

        for (State s : states) {
            s.abstractedBy = null;
        }

        restoreResistorIntoCircuit();

        root.addStates.addAll(states);
        root.addComponents.addAll(resistors);

        root.removeProcess(this);
    }

    @Override
    public void simProcessFlush() {
        double current = (aPin.state - bPin.state) * getResistanceInverse();
        double voltage = aPin.state;
        Iterator<Resistor> ir = resistors.iterator();

        for (State s : states) {
            Resistor r = ir.next();
            voltage -= r.getResistance() * current;
            s.state = voltage;
        }
    }

    @Override
    public void addToSubsystem(SubSystem s) {
        s.addProcess(this);
        super.addToSubsystem(s);
    }

    @Override
    public void quitSubSystem() {
    }

    @Override
    public void dirty(Component component) {
        recalculateResistance();
        if (isAbstracted())
            abstractedBy.dirty(this);
    }

    @Override
    public SubSystem getAbstractorSubSystem() {
        return getSubSystem();
    }
}

File: src/main/java/mods/eln/sim/mna/component/Line.kt
package mods.eln.sim.mna.component

import mods.eln.sim.mna.RootSystem
import mods.eln.sim.mna.SubSystem
import mods.eln.sim.mna.misc.ISubSystemProcessFlush
import mods.eln.sim.mna.state.State
import java.util.LinkedList

class Line : Resistor(), ISubSystemProcessFlush, IAbstractor {

    var resistors = LinkedList<Resistor>() //from a to b
    var states = LinkedList<State>() //from a to b

    override var abstractorSubSystem: SubSystem
        get() = subSystem ?: throw Exception()
        set(value) {}

    internal var ofInterSystem: Boolean = false

    internal fun canAdd(c: Component): Boolean {
        return c is Resistor
    }

    internal fun add(c: Resistor) {
        ofInterSystem = ofInterSystem or c.canBeReplacedByInterSystem()
        resistors.add(c)
    }

    override fun canBeReplacedByInterSystem(): Boolean {
        return ofInterSystem
    }

    fun recalculateR() {
        var R = 0.0
        for (r in resistors) {
            R += r.r
        }
        r = R
    }

    internal fun restoreResistorIntoCircuit() {
        this.breakConnection()
    }

    internal fun removeResistorFromCircuit() {}

    override fun returnToRootSystem(root: RootSystem?) {
        for (r in resistors) {
            r.abstractedBy = null
        }

        for (s in states) {
            s.abstractedBy = null
        }

        restoreResistorIntoCircuit()

        root!!.addStates.addAll(states)
        root.addComponents.addAll(resistors)

        root.removeProcess(this)
    }

    override fun simProcessFlush() {
        val i = (aPin!!.state - bPin!!.state) * rInv
        var u = aPin!!.state
        val ir = resistors.iterator()
        val `is` = states.iterator()

        while (`is`.hasNext()) {
            val s = `is`.next()
            val r = ir.next()
            u -= r.r * i
            s.state = u
        }
    }

    override fun addedTo(s: SubSystem) {
        s.addProcess(this)
        super.addedTo(s)
    }

    override fun quitSubSystem() {}

    override fun dirty(component: Component) {
        recalculateR()
        abstractedBy?.dirty(this)
    }

    companion object {
        fun newLine(root: RootSystem, resistors: LinkedList<Resistor>, states: LinkedList<State>) {
            if (resistors.isEmpty()) {
            } else if (resistors.size == 1) {
            } else {
                val first = resistors.first
                val last = resistors.last
                val stateBefore = if (first.aPin === states.first) first.bPin else first.aPin
                val stateAfter = if (last.aPin === states.last) last.bPin else last.aPin

                val l = Line()
                l.resistors = resistors
                l.states = states
                l.recalculateR()
                root.addComponents.removeAll(resistors)
                root.addStates.removeAll(states)
                root.addComponents.add(l)
                l.connectTo(stateBefore, stateAfter)
                l.removeResistorFromCircuit()

                root.addProcess(l)

                for (r in resistors) {
                    r.abstractedBy = l
                    l.ofInterSystem = l.ofInterSystem or r.canBeReplacedByInterSystem()
                }

                for (s in states) {
                    s.abstractedBy = l
                }
            }
        }
    }
}
